Karnage/histogram

GDBobby opened the pull request:
generates 1d perlin noise and converts it to a bmp image of a histogram. 
size of image is controllable via command line, from 256 to 4096, in multiples of 256

karnkaul committed:
Add examples/histogram

karnkaul committed:
Add build CI.

karnkaul committed:
Fix tests, update CMake presets.

karnkaul committed:
Fix non-constexpr function.

GDBobby committed:
Merge branch 'main' into karnage/histogram

karnkaul commented:
Uh, I don't see any changes 😅

GDBobby committed:
histogram bitmap

the changes

GDBobby committed:
fixing errors

idk how to run the linux test locally

GDBobby committed:
error checkign 2

should cover it?

GDBobby committed:
error checking 3

reserve requires size_t???

GDBobby committed:
error checking 4

signedness

GDBobby committed:
doesnt know what a histogram is

no more interpolation between points, now correctly creating a bar graph

fixed, now correctly has
assert <= 16
instead of
assert < 16

GDBobby commented:
ok hopefully its good now

karnkaul commented on examples/HistogramBitmap/CMakeLists.txt:
Generally the convention for user-facing binaries is either `kebab-case` or `PascalCase`, `noiz-HistogramBitmap` is a mix of both.

karnkaul commented on examples/HistogramBitmap/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+)
+
+target_sources(${PROJECT_NAME} PRIVATE
+  HistogramBitmap.cpp
```
I prefer to stick to `snake_case` for file paths because different OSs have different path case sensitivities.

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
Looks good overall! Needs some small changes:
- Run `clang-format` (Format Document) on the file, this can be done automatically on save if you like (in VSCode).
- Fix the linter warnings (install the `clang-tidy` plugin if you don't see any warnings).

Screenshot as example of linter issues (all the stuff highlighted green, as well as the hover window on a specific one):

![image](https://github.com/cpp-gamedev/noiz/assets/16272243/246a8f49-7481-463c-b7fb-c6bf70c2728d)

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+#include <vector>
+#include <fstream>
+
+	struct BmpHeader {
```
Alternatively, you could use PPM as the output format instead, much easier to deal with.

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+
+
+namespace {
+template <typename Type>
+concept NumberT = std::integral<Type> || std::floating_point<Type>;
+
+template <NumberT Type>
+auto parse_as(Type& out, std::string_view const value) -> bool {
+	if (value.empty()) { return false; }
+
+	auto const* end = value.data() + value.size();
+	auto const [ptr, ec] = std::from_chars(value.data(), end, out);
+
+	return ec == std::errc{} && ptr == end;
+}
+
+struct Args {
+	std::span<char const* const> args{};
+
+	[[nodiscard]] constexpr auto next() -> std::string_view {
+		if (args.empty()) { return {}; }
+		auto const* ret = args.front();
+		args = args.subspan(1);
+		return ret;
+	}
+
+	template <NumberT Type>
+	auto next_as(Type& out, std::string_view const key) -> bool {
+		auto value = next();
+		if (value.empty()) { return true; }
+
+		if (!parse_as(out, value)) {
+			std::cerr << std::format("invalid argument: '{}' for '{}'\n", value, key);
+			return false;
+		}
+
+		return true;
+	}
+};
```
Technically this is repeated code, but since it's supposed to be a self-contained example it's fine. Perhaps if we add more examples, we could add a common header with this stuff that's shared among all of them. Thoughts?

GDBobby replied:
For sure. Is there a metaprogramming style to iterate through every member of an unknown object? Never been able to figure it out before.

karnkaul replied:
Not yet, that will require reflection (C++26 earliest).

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+		if (!args.next_as(count, "count")) { return false; }
+		if (!args.next_as(step, "step")) { return false; }
+		if (!args.next_as(imageSize, "imageSize")) { return false; }
+		assert(imageSize > 0 && imageSize <= 16);
```
Don't `assert` about an input that could be anything, it will unnecessarily crash the program. Instead print the error and return false. Similarly, the "unrecognized argument" part is handled here, so it doesn't need to be printed again in main.

```suggestion
		if (imageSize < 0 || imageSize > 16) {
		    std::cerr << std::format("invalid image size: '{}'\n", imageSize);
		    return false;
		}
```

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+// contains logic to store data points and draw them
+class HistogramBitmap {
+  public:
+	struct Color{
```
`Rgb` might be a better name, since `Color` is a bit ambiguous. (Could be RGB / RGBA / CMYK / hex code / etc.)

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+			width = imageSize;
+			height = imageSize;
+		}
+		void writeToFile(std::ofstream& outFile){
```
Member function should be `const`.

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+		uint32_t colorTableEntries = 0;
+		uint32_t importantColors = 0;
+
+		BmpInfoHeader(uint16_t imageSize){
+			width = imageSize;
+			height = imageSize;
+		}
```
Use member list initialization instead of assignment in the constructor body.

```suggestion
		BmpInfoHeader(uint16_t const imageSize) : width(imageSize), height(imageSize)
```

karnkaul commented on examples/HistogramBitmap/HistogramBitmap.cpp:
```diff
@@ -0,0 +1,234 @@
+		return *this;
+	}
+
+	void writeImage() {
```
Member function should be `const`.

karnkaul reviewed (changes requested):
Nice work! Some changes needed before it's good to be merged.

I noticed that this file uses `camelCase` where none of the others in the project do. It's not a big deal, but it may confuse readers a bit.

GDBobby committed:
best practice

swapping formatting of variables to match other project variables,
few data name changes to more accurately reflect what the data does

global function used to write bitmap header swapped to a static structure member function, of BMP_Header
combined both bmp header structures into one

GDBobby committed:
implicit casting fix

size of bmp file
image_size is a uint16_t, which would remove the need to static cast, but I believe if image_size is 4096. There's potential for overflow, again reintroducing the need to cast up to uint32_t. Not really sure what the code will do behind the scenes, so probably best to use the more readable width * height

GDBobby commented:
Tempted to not allow the construction of the BMP header struct outside of the static writing function.
Hopefully fine now.

GDBobby committed:
minor error fix

not correctly taking the difference between exactPoint and flooredValue leads to a shift in the bars of the histogram

GDBobby commented:
usage covered by example is caught in a cross section between the plaintext histogram, and the texture generator found in another example.
Closing branch
